cdap-client/src/main/java/co/cask/cdap/client/MetadataClient.java

```diff
@@ -0,0 +1,473 @@
+import co.cask.common.http.HttpResponse;
+import com.google.common.reflect.TypeToken;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+
+import java.io.IOException;
+import java.lang.reflect.Type;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.util.Map;
+import java.util.Set;
+import javax.annotation.Nullable;
+import javax.inject.Inject;
+
+/**
+ * Provides ways to interact with CDAP Metadata.
+ */
+public class MetadataClient {
+
+  private static final Gson GSON = new GsonBuilder()
+    .registerTypeAdapter(Id.NamespacedId.class, new NamespacedIdCodec()).create();
+  private static final Type SET_METADATA_RECORD_TYPE = new TypeToken<Set<MetadataRecord>>() { }.getType();
+  private static final Type SET_METADATA_SEARCH_RESULT_TYPE =
+    new TypeToken<Set<MetadataSearchResultRecord>>() { }.getType();
+
+  private final RESTClient restClient;
+  private final ClientConfig config;
+
+  @Inject
+  public MetadataClient(ClientConfig config, RESTClient restClient) {
+    this.config = config;
+    this.restClient = restClient;
+  }
+
+  public MetadataClient(ClientConfig config) {
+    this.config = config;
+    this.restClient = new RESTClient(config);
+  }
+
+  /**
+   * Returns search results for metadata.
+   *
+   * @param namespace the namespace to search in
+   * @param query the query string with which to search
+   * @param target the target type. If null, all possible types will be searched
+   * @return A set of {@link MetadataSearchResultRecord} for the given query.
+   */
+  public Set<MetadataSearchResultRecord> searchMetadata(Id.Namespace namespace, String query,
+                                                        @Nullable MetadataSearchTargetType target)
+    throws IOException, UnauthorizedException {
+
+    String path = String.format("metadata/search?query=%s", query);
+    if (target != null) {
+      path = path + "&target=" + target.getInternalName();
+    }
+    URL searchURL = config.resolveNamespacedURLV3(namespace, path);
+    HttpResponse response = restClient.execute(HttpRequest.get(searchURL).build(),
+                                               config.getAccessToken());
+    return GSON.fromJson(response.getResponseBodyAsString(), SET_METADATA_SEARCH_RESULT_TYPE);
+  }
+
+
+  /**
+   * @param appId the app for which to retrieve metadata
+   * @return The metadata for the application.
+   */
+  public Set<MetadataRecord> getMetadata(Id.Application appId)
+    throws IOException, UnauthorizedException, NotFoundException, BadRequestException {
+    String path = String.format("apps/%s/metadata", appId.getId());
+    return getMetadata(appId, path);
+  }
+
+  /**
+   * @param datasetInstance the dataset for which to retrieve metadata
+   * @return The metadata for the dataset.
+   */
+  public Set<MetadataRecord> getMetadata(Id.DatasetInstance datasetInstance)
+    throws IOException, UnauthorizedException, NotFoundException, BadRequestException {
```

Comment:
Hey @jwang47, that doesn't reduce the number of removeProperty methods, right?
It just replaces the constructPath methods into a map containing several functions and one constructPath method, right?

How is that better than having the functions be their own methods, and having the java compiler match the Id type to method (by the method's signature) rather than have a map of all the functions and do the mapping ourself?

For instance, in the PR, I have this:
```java
  private String constructPath(Id.Application appId) {
    return String.format("apps/%s", appId.getId());
  }

  private String constructPath(Id.Program programId) {
    return String.format("apps/%s/%s/%s",
                         programId.getApplicationId(), programId.getType().getCategoryName(), programId.getId());
  }

  private String constructPath(Id.Run runId) {
    Id.Program programId = runId.getProgram();
    return String.format("apps/%s/%s/%s/runs/%s",
                         programId.getApplicationId(), programId.getType().getCategoryName(), programId.getId(),
                         runId.getId());
  }

  private String constructPath(Id.DatasetInstance datasetInstance) {
    return String.format("datasets/%s", datasetInstance.getId());
  }

  private String constructPath(Id.Stream streamId) {
    return String.format("streams/%s", streamId.getId());
  }
```